designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+
+    // Return the wrapper.
+    return async function* formatter(resultIterator) {
+      // Collect and sort the results.
+      const results = []
+      for await (const result of resultIterator) {
+        results.push(result)
+      }
+      results.sort(ESLint.compareResultsByFilePath)
+
+      // Make `rulesMeta`.
+      const rules = this.cliEngine.getRules()
+      const rulesMeta = getRulesMeta(rules)
+
+      // Format the results with the formatter of the current spec.
+      yield format(results, { rulesMeta })
+    }
+  }
+}
+```
+
+</details>
+
+<details>
+<summary>Example: Use the formatter.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+const results = eslint.executeOnFiles(patterns)
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+</details>
+
+Once the `getFormatter()` method got this change, we can update the specification of custom formatters without breakage in the future to support streaming.
+
+#### § The other methods
+
+The following methods receive an `AsyncIterable<LintResult>` object as the first argument then return an `AsyncIterableIterator<LintResult>` object because those methods are sandwiched between `executeOnFiles()` and formatters.
+
+- `outputFixes()`
+- `getErrorResults()`
+
+<details>
+<summary>Example: Use `outputFixes()` and `getErrorResults()`.</summary>
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+const formatter = eslint.getFormatter("stylish")
+
+// Verify files
+let results = eslint.executeOnFiles(patterns)
+// Update the files of the results if needed
+// (This must be done before `--quiet`)
+if (process.argv.includes("--fix")) {
+    results = ESLint.outputFixes(results)
+}
+// Filter the results if needed
+if (process.argv.includes("--quiet")) {
+    results = ESLint.getErrorResults(results)
+}
+// Format and write the results
+for await (const textPiece of formatter(results)) {
+    process.stdout.write(textPiece)
+}
+```
+
+</details>
+
+The following methods return `Promise` which gets fulfilled with each result in order to support plugins/configs that are ES modules without breakage in the future.
+
+- `getConfigForFile()`
+- `isPathIgnored()`
+
+The following method is as-is because those don't touch both file system and module system.
+
+- `getRules()`
+
+The following methods are removed because those don't fit the current API.
+
+- `addPlugin()` ... This method has caused to confuse people. We have introduced this method to add plugin implementations and expected people to use this method to test plugins, but people have thought that this method loads a new plugin for the following linting. And this method is only one that mutates the state of `CLIEngine` objects (except cache). This RFC moves this functionality to the constructor options.
```

Comment:
So main purpose is to remove any mutations of the state of `ESLint`? With bonus effect in avoiding misunderstandings in how it can be used?